unix: Write to UV_HANDLE_BLOCKING_WRITES streams in the thread pool

xal-0 opened the pull request:
On Unix, if a `uv_stream_t` has `UV_HANDLE_BLOCKING_WRITES` set, either because we were unable to put the file in `O_NONBLOCK` mode when the stream was initialized, or because the stream has been shared with a subprocess via `UV_INHERIT_STREAM`, `uv_write` can block.

As discussed in #3602, the nicest solution would be to use platform-specific features to do non-blocking writes to files regardless of the `O_NONBLOCK` state, but a fallback is needed regardless.  This change implements the fallback by doing all writes to blocking streams in the thread pool.

There are a handful of specific changes I'd like to get feedback on:
- This tweaks uv_spawn slightly: we clear `O_NONBLOCK` on libuv-owned streams shared using `UV_INHERIT_STREAM`, but not on files shared with `UV_INHERIT_FD`.  We set `UV_HANDLE_BLOCKING_WRITES` only when we put the file in blocking mode, since `uv_stream_set_blocking` is documented to make `uv_write` block.
- The `UV_HANDLE_*` enum is getting crowded.  I put `UV_HANDLE_WRITE_PENDING` on top of a Windows-only flag, but the name is generic enough that we might one day want to use it for Windows too.  Should it go somewhere else?
- `uv_fs_t* blocked_write` is a new field on `UV_STREAM_PRIVATE_FIELDS` for now, breaking ABI.

Handling `uv_close` is more complicated when we dispatch writes to the thread pool.  We can try to `uv_cancel` the write request, but there are few portable things we can do to cancel an in-progress write once it has been picked up from the work queue, save for reserving a signal for libuv's use only.  I implemented a compromise: we try `uv_cancel`, but keep the FD open and defer calling the write callbacks until the `write(2)` call completes (the write request itself doesn't need to be completed fully).  This seems acceptable because until #4966 the callback can't know how many bytes were written if `status == UV_ECANCELED`.

The added test reproduces a problem we encounter frequently in Julia, where spawning a process puts our stream into blocking mode, then a later write fills the pipe/TTY buffer and deadlocks.

xal-0 committed:
Write to UV_HANDLE_BLOCKING_WRITES streams in the threadpool

On Unix, if a uv_stream_t has UV_HANDLE_BLOCKING_WRITES set, either
because we were unable to put the file in O_NONBLOCK mode when the
stream was initialized, or because the stream has been shared with a
subprocess via UV_INHERIT_STREAM, uv_write can block.

The nicest solution would be to use platform-specific features to do
non-blocking writes to files regardless of the O_NONBLOCK state, but a
fallback is needed regardless.  This change implements the fallback by
doing all writes to blocking streams in the threadpool.

Note that this also tweaks uv_spawn slightly: we clear O_NONBLOCK on
libuv-owned streams shared using UV_INHERIT_STREAM, but not on files
shared with UV_INHERIT_FD.

Keno commented:
> We can try to `uv_cancel` the write request, but there are few portable things we can do to cancel an in-progress write once it has been picked up from the work queue, save for reserving a signal for libuv's use only

I do think we need a way to kick the thread out of the syscall for cancellation purposes. We already have `UV_LOOP_BLOCK_SIGNAL` - no reason we can't have `UV_LOOP_CANCEL_SIGNAL` have the embedder designate a signal that has been appropriately configured. Needs to properly handle the sigmask of course.

xal-0 committed:
Add pipe_blocking_cancel to test cancelling blocking write on close

xal-0 commented:
A summary of what happens when a pipe that is being written to on another thread is closed.  "End" is `r`/`w` depending on whether I closed the read or write end of the pipe. "(last)" indicates no other open fd refers to that end of the pipe.
| OS      | End      | Return | errno | Signal? |
|---------|----------|--------|-------|---------|
| Linux   | r (last) | >= 0   |       | SIGPIPE |
|         | r        | hang   |       |         |
|         | w (last) | hang   |       |         |
|         | w        | hang   |       |         |
| FreeBSD | r (last) | >= 0   |       |         |
|         | r        | hang   |       |         |
|         | w (last) | hang   |       |         |
|         | w        | hang   |       |         |
| macOS   | r (last) | -1     | EPIPE | SIGPIPE |
|         | r        | hang   |       |         |
|         | w (last) | -1     | EPIPE | SIGPIPE |
|         | w        | -1     | EPIPE | SIGPIPE |

Closing the last fd for the read end interrupts the `write(2)` everywhere, but FreeBSD doesn't give you SIGPIPE.  macOS seems to clean up the pipe if you close the final fd for either end, interrupting the write.  It is also the only platform where closing the fd being written to always results in EPIPE/SIGPIPE. I didn't try this with PTYs or TCP streams.

I was hoping we might be able to get the right behaviour by closing the fd for the stream, but this is a bit of a mess. I'll add the configurable libuv signal like @Keno suggested.

saghul commented:
> > IIRC way back when this was added so the console logs in Node wouldn't be lost (or maybe also interleaved?) if the process exits abruptly.
> 
> Do you know if they are using `uv_stream_set_blocking`? If so, we should still block in `uv_write`.

They are:

- https://github.com/nodejs/node/blob/2e597ded4bf8f90146647ab666123c3cb5069e78/src/stream_wrap.cc#L320 
- https://github.com/nodejs/node/blob/2e597ded4bf8f90146647ab666123c3cb5069e78/lib/tty.js#L111
- https://github.com/nodejs/node/blob/2e597ded4bf8f90146647ab666123c3cb5069e78/lib/net.js#L443

xal-0 committed:
Submit with uv__work_submit directly (no uv_fs_write)

xal-0 commented:
It's a work in progress, but I have a design working that uses `pthread_cancel` to interrupt the write() syscall in the thread pool. On unix, there are now a pair of functions, `uv__work_cancellable_enter()` and `uv__work_cancellable_leave()`, that flip the `PTHREAD_CANCEL_ENABLE` bit on the worker thread.  It defaults to disabled.

If we attempt to cancel work when it's not in one of these regions, `uv__work_cancel` returns `UV_BUSY` like before.  If it's in a cancellable region, the `done` callback will be called with `UV_ECANCELLED` after the thread is cancelled.  The watchdog thread monitors a queue of cancelled work items and restarts the worker thread that was cancelled.

I went with `pthread_cancel` because it's the only portable way, as far as I know, to solve the race between checking a flag set in a signal handler and entering an interruptable syscall; another small benefit is we don't need to reserve a signal like I thought we would.  It would of course be preferable to not have to restart cancelled threads like this, but it should be a rare enough occurrence for it not to matter.

This fix isn't completely satisfying, because it makes it impossible to read the return value from the interrupted syscall, which may have written some data before being interrupted.  That's ok right now, but won't be when we have `uv_write` cancellation.  Unfortunately I can think of only one fix, where we would make our own version of the "cancellable syscall" logic that Linux libcs use to implement `pthread_cancel`: We'd write the the flag check, the `write()` syscall instruction, and the return value check in assembly, then detect if the PC is inside the critical section in the signal handler.

xal-0 commented:
Yes, we need to be very careful to mark only regions where we call known cancellation points---the one I added surrounds `uv__try_write`, where `sendmsg`, `write`, and `writev` are the sole cancellation points.

xal-0 commented:
The ppc64 segfaults sure seem like bug in the version of libgcc Ubuntu 24.04's cross compiler ships with (13.3.0 I think?):
```
Thread 2 received signal SIGSEGV, Segmentation fault.
0x0000000010255b20 in uw_frame_state_for ()
(gdb) bt
#0  0x0000000010255b20 in uw_frame_state_for ()
#1  0x000000001025649c in _Unwind_ForcedUnwind_Phase2 ()
#2  0x0000000010256f00 in _Unwind_ForcedUnwind ()
#3  0x000000001017cc34 in __pthread_unwind ()
#4  0x0000000010170ac8 in sigcancel_handler ()
#5  0x00001000008107c8 in ?? ()
#6  0x0000000000000000 in ?? ()
```

I can't get it to happen on Debian 13, which ships 14.2.0.

I think I'd rather wire up `RWF_NOWAIT` than hack around this...

Keno commented:
> However, I don't know if POSIX guarantees this.

FWIW, (thought I had posted this), both freebsd and musl interpret POSIX as requiring this. Macos cancellation handling is more complicated and partially implemented in the kernel, but claude plausibly believes it has this behavior as well. I think we can assume it to be reliable.

xal-0 commented:
On second thought, this is an alternative that doesn't require an atomic check-flag-and-enter-syscall operation:
- Unmask cancellation signal
- Enter syscall
- If EINTR:
  - If `__thread volatile sigatomic_t was_cancelled` (set in the signal handler) true:
    - Atomic store to `work_done` and clean up
  - Else, ignore and retry syscall
- Else:
  - Atomic store to `work_done`

On the sending side, the protocol would be:
- Find thread doing work, removing it from the work queue if it has not yet started.
- While `work_done` is not true:
  - `pthread_kill` with the cancellation signal

I'm eager to avoid the finicky watchdog stuff that's needed for `pthread_cancel`, and I especially don't want to be doing the stack unwinding if we can avoid it.  If it sounds better and is sound I can implement this instead.

xal-0 committed:
Make pipe_blocking_* tests Unix-only

xal-0 committed:
unix: Add the ability to cancel thread pool work with a signal

This commit adds a new uv__work_kind, UV__WORK_FAST_IO_CANCELLABLE,
which allows us to uv_cancel work in the thread pool that has already
started executing.  It does this by signaling the thread with
pthread_kill, using a signal configurable with the UV_LOOP_CANCEL_SIGNAL
option.  Since we need to install a no-op signal handler, this signal
must be owned by libuv.  This could be relaxed, provided we document
that the signal must not have SA_RESTART set, and running the signal
handler on libuv worker threads must be harmless.

The uv__work struct gets two new fields on Unix: the thread that is
currently performing the work, and an atomic state field.  If the work
is not cancellable, it has state UV__WORK_BUSY the entire time.

If the work is cancellable, it is inserted into the work queue with
state UV__WORK_CANCELLABLE.  When we want to cancel the work, we first
check if the work is still in the queue (holding the queue lock), and if
it is not, we atomically check for UV__WORK_CANCELLABLE and swap in
UV__WORK_CANCEL_PENDING.  If that check succeeded, we signal the thread
with pthread_kill (the signal handler does nothing; we need only kick
the thread out of an interruptable syscall with EINTR).

In the worker thread, we check for UV__WORK_CANCEL_PENDING before and
after the syscall, setting it to UV__WORK_CANCELLED if we avoided
running the syscall, or if we got EINTR as the return value.  Since
there is a race between checking the flag and entering the syscall, we
prevent deadlocks by spinning on pthread_kill in uv__work_cancel, with
exponential backoff.  We wait a maximum of 1024 ms before returning
UV_EBUSY.

xal-0 commented:
While I still think `pthread_cancel` is the only way to cancel blocking syscalls while avoiding the race, I no longer think it's the best solution for this problem.  I tried something like https://github.com/libuv/libuv/pull/4976#issuecomment-3678185853, but still saw the `pthread_cancel` bugginess with mips64 glibc---it just isn't worth it to start debugging stack unwinding problems when we don't have a reason to unwind the stack in the first place.

This new implementation of cancellation is simpler and follows https://github.com/libuv/libuv/pull/4976#issuecomment-3676913638, more or less.  We still submit writes to the thread pool, but cancel them by setting an atomic flag on the work item and sending the specific thread a signal to kick it out of the interruptable syscall (see the commit message for details on the algorithm).  We loop, with exponential backoff, until the cancellation is acknowledged; after 1024 ms we give up.  Only `uv__write_work` can be cancelled right now.

saghul commented:
Coming back to this, AFAIS we'd be breaking the contract for the setp_blocking function. Currently `uv_write` blocks in that case, and Node IIRC relies on that.

xal-0 committed:
Fix the exponential backoff loop when cancelling work, return value

xal-0 committed:
Don't send handle twice when doing the write on a thread

xal-0 commented:
The last few commits address a few issues spotted in review, and it's all rebased on top of v1.x now.

Addressing @saghul's comment: The behaviour after calling `uv_stream_set_blocking` is unchanged (that is, it sets or clears `O_NONBLOCK` on the underlying file, but doesn't change how libuv uses it at all).  A stream that has been set to blocking with this function will block in `uv_write`, like before.  That function doesn't touch the `UV_HANDLE_BLOCKING_WRITES` flag on the stream on Unix.

We interpret `UV_HANDLE_BLOCKING_WRITES` to mean that the file is in blocking mode and it is outside of our control to change that.  Before this change, we set it on Unix only in `uv_tty_init` after failing to reopen the tty in order to put it in non-blocking mode.

After, we also set it in one more situation: when `uv__process_init_stdio` is setting up `UV_INHERIT_STREAM` stdios, we put the file for those streams into blocking mode and enable this flag, since there is no way for us to share a file in general without also sharing the non-blocking-ness.  This fixes the deadlocks we were seeing after spawning a process with libuv streams for stdios.

saghul commented:
I see, thanks for clarifying. I'll take another look.

vtjnash commented on include/uv/threadpool.h:
```diff
@@ -27,11 +27,23 @@
+#ifndef _WIN32
+  pthread_t thread;
+  _Atomic char state;
+#endif
```
This seems to be an ABI-break for many structs. Do we need to limit this bugfix to only v2 users?

xal-0 replied:
I've used one of the reserved pointers in the handle struct to work around this. The first time you make a write to a stream that may block, we allocate a `uv__blocked_write` struct and put it there. The cancellation protocol had to change a bit, since we can't rely on doing as much under the thread pool work queue lock, but it isn't a major change: we needed another state, called `WAITING`, which we release-CAS to `CANCELLABLE` after writing `pthread_self()` to the `uv__blocked_write` `thread` field.

vtjnash commented on src/unix/core.c:
```diff
@@ -344,7 +344,12 @@ static void uv__finish_close(uv_handle_t* handle) {
-      uv__stream_destroy((uv_stream_t*)handle);
+      if (handle->flags & UV_HANDLE_WRITE_PENDING) {
+        handle->flags ^= UV_HANDLE_CLOSED;
+        uv__make_close_pending(handle); /* Back into the queue. */
```
The writecb should put it back
```suggestion
```

vtjnash commented on src/unix/loop.c:
```diff
@@ -227,13 +230,23 @@ int uv__loop_configure(uv_loop_t* loop, uv_loop_option option, va_list ap) {
+    memset(&sa, 0, sizeof sa);
+    sa.sa_handler = &uv__cancel_signal_handler;
+    if (sigaction(loop->cancel_signum, &sa, NULL) == -1) {
+      loop->cancel_signum = -1;
```
Can we make this more-global, since this is per-loop config, but it affects all loops (and we should uninstall it back to SIG_DFL when the last loop is destroyed)

xal-0 replied:
I've made this a global config option set with `uv_threadpool_set_cancel_signal`. Since we don't have any requirements on the signal handler, other than it not be `SA_RESTART`, we document this and require the user set up a signal handler (that they can use for other purposes, too, if they want). That way we don't need to do reference counting or anything like that.

vtjnash commented on src/threadpool.c:
```diff
@@ -303,8 +348,38 @@ static int uv__work_cancel(uv_loop_t* loop, uv_req_t* req, struct uv__work* w) {
+      thread = w->thread;
+      for (i = 0; i < 10; ++i) {
+        if (pthread_kill(thread, loop->cancel_signum) != 0)
+          abort();
```
Should we make this reliable on kernels where that is possible (e.g. anything but linux), e.g. with `thread_abort(pthread_mach_thread_np(thread))` or with `thr_wake(thread)` on freebsd (which doesn't even need the retry loop, since it is actually guaranteed to be reliable by design)

vtjnash commented:
> The behaviour after calling uv_stream_set_blocking is unchanged (that is, it sets or clears O_NONBLOCK on the underlying file, but doesn't change how libuv uses it at all) 

Ah, this seems the main confusion then, as Windows and Linux used the same flag to mean nearly opposite things. Makes sense now that the current implementation is a bugfix for how this flag is used in tty on unix.

xal-0 committed:
Finish closing stream in uv__write_done instead of spinning

xal-0 committed:
Make uv_write_t's result field an `int`, like before

This field didn't actually need to be an `ssize_t`, so we can make it an `int`
to avoid changing the struct size.

xal-0 committed:
Don't watch for POLLOUT when stream has pending blocking write

On Unix, when there is a pending write to a stream with
UV_HANDLE_BLOCKING_WRITES set, we sholdn't watch for POLLOUT.  If we do,
and the stream becomes ready (but the write has not completed on the
threadpool worker), we'll repeatedly call uv__write and do nothing.
We're now more careful to set it only when we do a write on the
threadpool and get EAGAIN back.

The test for this issue was written with Claude Fable 5.  It catches the
problem pre-fix, and passes now.

xal-0 committed:
Make UV_LOOP_CANCEL_SIGNAL global, unmask in worker, document

Assisted-by: Claude:claude-fable-5

xal-0 committed:
Use plain char for uv__work.state, to avoid public _Atomic

Assisted-by: Claude:claude-fable-5

xal-0 committed:
Closing streams should not abort on write

Since closing a stream leaves the FD open until any pending writes have
been cancelled, we need to check that the stream isn't closing before
doing a write.

Assisted-by: Claude:claude-fable-5

xal-0 committed:
Unblock the cancel signal only when writing on thread pool

xal-0 committed:
Use reserved handle field to be ABI-compatible

xal-0 committed:
Don't sleep in uv__write_cancel if no cancel signal configured

xal-0 committed:
Check uv__is_closing, in addition to fd == -1

xal-0 commented:
These last few commits resolve some issues that came up during LLM review, as well as @vtjnash's suggestions.  The big change is I've ported everything to be ABI compatible with v1 now, and I've abandoned the effort to make any of the thread pool cancellation stuff reusable for other work items.  The cancellation signal is also more tightly scoped: we unblock the cancellation signal just before entering the write syscall, and re-block it after.  There are some explanatory comments explaining the `uv__blocked_write` state transitions.

vtjnash commented on test/test-pipe-blocking-subprocess.c:
```diff
@@ -150,10 +150,19 @@ TEST_IMPL(pipe_blocking_subprocess) {
 static void timer_cb(uv_timer_t *handle) {
+  uv_write_t after_close_req;
+
   /* The write has begun by now.  We'll close the write side of the pipe first,
    * since closing the read side will trigger SIGPIPE. */
```
```suggestion
```
Comment no longer makes sense, since `uv_close` does not impose an ordering (and never did on Windows anyways)

vtjnash commented on test/test-list.h:
```diff
@@ -772,6 +770,9 @@ TASK_LIST_START
   TEST_ENTRY  (tcp_listen_without_bind)
+  TEST_ENTRY_CUSTOM (pipe_blocking_cancel, 0, 0, 20000)
+  TEST_ENTRY_CUSTOM (pipe_blocking_no_busy_poll, 0, 0, 20000)
+  TEST_ENTRY_CUSTOM (pipe_blocking_subprocess, 0, 0, 20000)
```
20 seconds seems pretty slow. Do they all usually take a long time, or could we use the default timeout (250 ms, IIRC)

xal-0 replied:
I'll try these with the default timeout and see if it causes any problems on CI.

vtjnash commented on src/unix/stream.c:
```diff
@@ -802,14 +888,24 @@ static int uv__try_write(uv_stream_t* stream,
-    while (n == -1 && errno == EINTR);
+    } while (n == -1 && errno == EINTR);
   }
+  if (w != NULL)
+    uv__block_cancel(1);
```
The necessity to block the signal seems a bit annoying, would it help to use the same cancellation protocol from win/pipe.c here (which replaces these syscalls with a lock):

https://github.com/libuv/libuv/blob/2cadaa40167050baf7c6905ac897e6fb57afb2c6/src/win/pipe.c#L1348-L1352

xal-0 replied:
As much as I'd like to get rid of the signal blocking/unblocking, I don't think we can't do the same cancellation protocol as on Windows, because `pthread_kill` is delivered asynchronously. On the cancelling thread, we could acquire the lock, send the signal, observe `state == UV__WRITE_CANCELLED`, and release the lock. The thread pool thread would finish the write, acquire/release the lock, and then receive the signal at some unexpected point in the future, while doing other work.

It's possible we could make it work by installing a signal handler and requiring the signal be acknowledged before we release the lock on the cancelling thread.  I don't like it as much since we'd have to override the signal handler for the whole process.  The way it is right now, the user need only provide a signal where executing the signal handler on a thread pool thread has no ill effects.

vtjnash replied:
There's a syscall for that (sigwait)

xal-0 replied:
We'd still have to block and unblock the signal, no?

vtjnash replied:
Ah, yes, there's a slight wrinkle that (if we got canceled) to make sigwait atomic with the delivery of the signal, we need to block the signal, check if the signal was received yet, call sigwait, then unblock the signal

vtjnash commented on src/unix/stream.c:
This looks like it would be a data-race now (loading the flags field of send_handle)--it may need to be done on the main thread before we launch the thread, then set a flag to disallow closing send_handle (seems similar to write-pending flag)

xal-0 replied:
I've thought about this more and tried to do this (defer closing if we're sending a handle on a blocking stream), but it's very brittle and I have little confidence in the implementation I landed on.  I also tried one where we immediately duplicate the fd on the loop thread and close it after writing it to the stream, which seems easier.

Would it be unreasonable to document that handles written to a stream must remain valid until the write callback is called?

xal-0 replied:
The most recent commit is an implementation of the "duplicate FD before sending on thread pool" idea, since it's actually pretty simple.

xal-0 committed:
Use the default timeout for pipe_blocking_* tests

xal-0 committed:
Merge remote-tracking branch 'upstream/v1.x' into ss/stream-write-threadpool

xal-0 committed:
Test cancelled partial writes on thread pool

Assisted-by: Codex:gpt-5.6-sol

xal-0 commented:
The most recent merge here has some tricky code, now that we can cancel writes and need to do callbacks with the partial progress. The commit after that contains a vibe coded test for one of the tricky cases.

xal-0 committed:
Duplicate send_handle fd before sending on thread pool

xal-0 committed:
Immediately close the fd for streams without pending writes
